arch/arm/core/aarch64/reset.S

```diff
@@ -44,60 +49,89 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__reset)
 GTEXT(__start)
 SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
 
-#ifdef CONFIG_SWITCH_TO_EL1
-	switch_el x1, 3f, 2f, 1f
-3:
-	/* Disable MMU and async exceptions routing to EL1 */
-	msr	sctlr_el1, xzr
-
-	/* Disable EA/IRQ/FIQ routing to EL3 and set EL1 to AArch64 */
-	mov	x0, xzr
-	orr	x0, x0, #(SCR_EL3_RW)
-	msr	scr_el3, x0
-
-	/* On eret return to EL1 with DAIF masked */
-	mov	x0, xzr
-	orr	x0, x0, #(DAIF_MASK)
-	orr	x0, x0, #(SPSR_EL3_TO_EL1)
-	orr	x0, x0, #(SPSR_EL3_h)
-	msr	spsr_el3, x0
-
-	adr	x0, 1f
-	msr	elr_el3, x0
-	eret
-2:
-	/* Boot from EL2 not supported */
-	bl	.
-1:
-#endif
 	/* Setup vector table */
-	adr	x0, _vector_table
+	adr	x9, _vector_table
 
 	switch_el x1, 3f, 2f, 1f
 3:
 	/* Initialize VBAR */
-	msr	vbar_el3, x0
+	msr	vbar_el3, x9
+	isb
+
+	/* Initialize sctlr_el3 to reset value */
+	mov_imm	x2, SCTLR_EL3_RES1
+	mrs     x1, sctlr_el3
+	orr	x1, x1, x2
+	msr	sctlr_el3, x1
+	isb
 
 	/* SError, IRQ and FIQ routing enablement in EL3 */
-	mrs	x0, scr_el3
-	orr	x0, x0, #(SCR_EL3_IRQ | SCR_EL3_FIQ | SCR_EL3_EA)
-	msr	scr_el3, x0
+	mrs	x1, scr_el3
+	orr	x1, x1, #(SCR_EL3_IRQ | SCR_EL3_FIQ | SCR_EL3_EA)
+	msr	scr_el3, x1
+
+	/*
+	* Disable access traps to EL3 for CPACR, Trace, FP, ASIMD,
+	* SVE from lower EL.
+	*/
+	mov_imm	x1, CPTR_EL3_RES_VAL
+	mov_imm	x2, (CPTR_EL3_TTA | CPTR_EL3_TFP | CPTR_EL3_TCPAC)
+	bic	x1, x1, x2
+	orr	x1, x1, #(CPTR_EL3_EZ)
+	msr	cptr_el3, x1
+	isb
+
+	/* Platform specific configurations needed in EL3 */
+	bl	z_arch_el3_plat_init
 
-	/* Disable access trapping in EL3 for NEON/FP */
-	msr	cptr_el3, xzr
+#ifdef CONFIG_SWITCH_TO_EL1
+	/*
+	* Zephyr entry happened in EL3. Do EL3 specific init before
+	* dropping to lower EL.
+	*/
+	/* Enable access control configuration from lower EL */
+	mrs	x1, actlr_el3
+	orr     x1, x1, #(ACTLR_EL3_L2ACTLR | ACTLR_EL3_L2ECTLR \
+			 | ACTLR_EL3_L2CTLR)
+	orr     x1, x1, #(ACTLR_EL3_CPUACTLR | ACTLR_EL3_CPUECTLR)
+	msr	actlr_el3, x1
+
+	/* Initialize sctlr_el1 to reset value */
+	mrs     x1, sctlr_el1
+	mov_imm	x2, SCTLR_EL1_RES1
+	orr	x1, x1, x2
+	msr	sctlr_el1, x1
 
+	/* Disable EA/IRQ/FIQ routing to EL3 and set EL1 to AArch64 */
+	mov	x1, xzr
+	orr	x1, x1, #(SCR_EL3_RW)
+	msr	scr_el3, x1
+
+	/* On eret return to secure EL1h with DAIF masked */
+	mov	x1, xzr
+	orr	x1, x1, #(DAIF_MASK)
+	orr	x1, x1, #(SPSR_EL3_TO_EL1)
+	orr	x1, x1, #(SPSR_EL3_h)
+	msr	spsr_el3, x1
+
+	adr	x1, 1f
+	msr	elr_el3, x1
+	eret
+#endif
 	/*
 	 * Enable the instruction cache, stack pointer and data access
-	 * alignment checks and disable speculative loads.
+	 * alignment checks.
 	 */
-	mov	x1, #(SCTLR_I_BIT | SCTLR_A_BIT | SCTLR_SA_BIT)
-	mrs	x0, sctlr_el3
-	orr	x0, x0, x1
-	msr	sctlr_el3, x0
+	mov	x2, #(SCTLR_I_BIT | SCTLR_A_BIT | SCTLR_SA_BIT)
+	mrs	x1, sctlr_el3
+	orr	x1, x1, x2
+	msr	sctlr_el3, x1
```

Comment:
Same here, reusing the same registers helps in keeping the diff smaller.

Reply:
ACK